File: python/gpa.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

'''
    Last reviewed: 26/10/2019
    author: G. A. Moreira
'''

import scipy
import numpy as np
import matplotlib.pyplot as plt

from math import radians

from mpl_toolkits import mplot3d
from sklearn.neighbors import NearestNeighbors


def rotmat(alpha, beta, gamma):
    '''
        Builds a numpy.ndarray rotation matrix with the provided rotation
        angles, (in degrees). Gamma is the angle about the x-axis, beta about
        the y-axis and alpha about the z-axis.
    '''
    alpha = radians(alpha)
    beta = radians(beta)
    gamma = radians(gamma)

    Rx = np.array([[1, 0, 0],
                   [0, np.cos(gamma), -np.sin(gamma)],
                   [0, np.sin(gamma), np.cos(gamma)]])

    Ry = np.array([[np.cos(beta), 0, np.sin(beta)],
                   [0, 1, 0],
                   [-np.sin(beta), 0, np.cos(beta)]])

    Rz = np.array([[np.cos(alpha), -np.sin(alpha), 0],
                   [np.sin(alpha), np.cos(alpha), 0],
                   [0, 0, 1]])

    R = np.matmul(Rz,Ry,Rx)
    
    return R



def mutual_nearest(clouds):
    '''
        Receives any number (N) of point clouds as an input. The point clouds 
        should be type numpy.ndarray with the first dimension being the points
        (P) and the second the coordinates (x,y,z) e.g.
        Outputs a numpy.ndarray of dimensions (Q,N). Q is the number of groups
        of mutually nearest neighbours and N are all the point clouds. This
        array contains the indices to select the points from each cloud. So 
        the 4th row of the array contains N indices to select the points from
        each cloud that form a mutually nearest neighbor group.
    '''    
    n_clouds = len(clouds)
    points_per_cloud = [cloud.shape[0] for cloud in clouds]
    n_points = np.sum(points_per_cloud)    
    correspondences = np.zeros([n_points, n_points])
    corr = np.zeros([1000, n_clouds, n_clouds], dtype='uint16')
    
    for i in range(n_clouds):
        for j in range(n_clouds):
            nbrs = NearestNeighbors(1, algorithm='ball_tree').fit(clouds[j])                                       
            _, indices_j = nbrs.kneighbors(clouds[i])
            corr[:, i, j] = indices_j.flatten()
        
    for i in range(n_clouds):
        for j in range(n_clouds):
            match = np.where(np.arange(0,corr.shape[0]) - corr[corr[:,i,j],j,i] == 0)
            index_i = corr[match,i,i].flatten() + i*points_per_cloud[i]
            index_j = corr[match,i,j].flatten() + j*points_per_cloud[j]
            correspondences[(index_i, index_j)] = 1
        
    return correspondences


def procrustes(A, B):
    '''
        Solves the Procrustes problem using the closed-form solution presented
        in Schonemann (1977). Given two (p,q) matrices A, B it finds the scalar
        S, the rotation matrix R and the translation vector T that minimize the
        error E = s*A*R + j*T' - B (i.e. the Frobenius norm of E). The solution
        involves the use of singular value decomposition (SVD) and is fully
        detailed in the article.
    '''
    
    centroid_B = np.mean(B, axis=0)
    centroid_A = np.mean(A, axis=0)
    
    A_centered = A - centroid_A
    B_centered = B - centroid_B
    
    C = np.matmul(A_centered.T, B_centered)
    
    V, _, W = np.linalg.svd(C)
    
    R = np.matmul(V, W).T
    detR = np.linalg.det(R)
    
    if detR < 0:
        reflex = np.diag([1,1,detR])
        R = np.matmul(np.matmul(V, reflex), W)
        
    T = centroid_B - np.matmul(R, centroid_A)
    s = 1
    
    return s, R, T  



def cloudviz(clouds, **kwargs):
    '''
        Receives a list of numpy.ndarray point clouds and represents them in a
        3D coordinate system.
    '''
    plt.figure(facecolor='0')
    ax = plt.axes(projection='3d')
    ax.set_facecolor('black')
    for i in range(len(clouds)):
        ax.plot(clouds[i][:,0],clouds[i][:,1],clouds[i][:,2], 'o', markersize=1)
    ax.legend(['Point cloud ' + str(i) for i in range(len(clouds))])
    plt.autoscale('off')
    plt.axis('off')
    plt.grid(b=None)
    plt.show()
    


def SRTtransform(cloud, s, R, T):
    '''
        Transforms a point cloud provided as a numpy.ndarray, according to a
        scale factor s, a rotation R and a translation T. The cloud should
        have dimensions (p,q) where p is the number of points and q is the 
        number of dimensions. s is a float, R a (3,3) numpy.ndarray and T is
        a (1,3) numpy.ndarray.
    '''
    rotated = np.matmul(R, cloud.T).T 
    scaled = s*rotated
    translated = scaled + T.reshape([1,3])

    return translated


class Centroid:
    '''
    '''
    def __init__(self, groups, stacked_clouds, points_per_cloud):
        '''
        '''
        self.centroid = []
        self.points_per_cloud = points_per_cloud
        self.groups = groups
        for i in range(len(self.groups)):
             points = stacked_clouds[self.groups[i]]
             mean = np.mean(points, axis=0)
             self.centroid.append(mean)
        
        self.centroid = np.array(self.centroid)


    def filterByCloud(self, cloud_number):
        '''
        '''
        start_index = np.sum(self.points_per_cloud[:cloud_number])
        end_index = np.sum(self.points_per_cloud[:1+cloud_number])
        cloud_points = []
        centroid_points = []
        for i in range(len(self.groups)):
            group = self.groups[i]
            condition = np.logical_and(group >= start_index, group < end_index)
            condition_true = np.where(condition)[0]
            if len(condition_true) == 1:
                centroid_points.append(self.centroid[i,:])
                cloud_points.append(stacked_clouds[group[condition_true[0]]])

        return np.array(cloud_points), np.array(centroid_points)
    
    
#%%


mat = scipy.io.loadmat('bunny.mat')
clouds_ = [mat['modelSimply'][0][i][0] for i in range(3)]



#%%
# Register the point clouds using Procrustes within a framework similar to ICP

clouds = clouds_.copy()

MAX_ITER = 50
N_CLOUDS = len(clouds)
POINTS_PER_CLOUD = [cloud.shape[0] for cloud in clouds]
    
for i in range(MAX_ITER):
    print('Iteration: ' + str(i))
    
    stacked_clouds = np.vstack(clouds)
    correspondence_map = mutual_nearest(clouds)
    vals, idx_start, count = np.unique(correspondence_map, axis=0, return_counts=True, return_index=True)
    points_per_group = np.sum(vals, axis=1)
    correspondence_map_unique = vals[np.logical_and(points_per_group==count, count>1)]
    
    group_id, point_indices = np.array(np.where(correspondence_map_unique==1), dtype='uint16')
    split_indices = np.unique(group_id, return_index=True)[1]
    groups = np.split(point_indices, split_indices)

    err_before = 0
    err_after = 0
    
    for k in range(N_CLOUDS):
        print('   Point cloud: ' + str(k))
        centroid = Centroid(groups, stacked_clouds, POINTS_PER_CLOUD)
        A, B = centroid.filterByCloud(k)
        print('      No. Points used (centroid): ' + str(A.shape[0]))
        
        cloud_error_before = np.sum(np.square(A-B))
        err_before += cloud_error_before
        print('      Cloud-Centroid MSE (before): ' + str(round(err_before/A.shape[0],3)))
        
        s, R, T = procrustes(A,B)
        clouds[k] = SRTtransform(clouds[k],s,R,T)

        cloud_error_after = np.sum(np.sum(np.square(SRTtransform(A,s,R,T)-B))) 
        err_after += cloud_error_after 
        print('      Cloud-Centroid MSE (after): ' + str(round(err_after/A.shape[0],3)))
        
    err_before /= stacked_clouds.shape[0]
    err_after /= stacked_clouds.shape[0]
    
    print('MSE before-after: ' + str(round(err_before,4)) +\
           str('  ' ) + str(round(err_after,4)))




    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    

File: src/gpa.cpp
// gpa.cpp
// Date: 2019-29-10
// Created by: Gabriel Moreira

#include <vector>
#include <map>
#include <math.h>
#include <Eigen/Dense>

#include <ANN/ANN.h>
#include "gpa.h"


float eigenMSE(Eigen::MatrixXf target, Eigen::MatrixXf current) {
  // Computes the mean-square error between two point matrices
  Eigen::MatrixXf delta;
  Eigen::MatrixXf ddt;
  float mse;

  delta = target - current;
  ddt = delta * delta.transpose();
  mse = ddt.trace() / delta.rows();
  return mse;
}


std::vector<float> eigenToVector(Eigen::MatrixXf mat) {
  // Converts Eigen::Matrix to std::vector<float> assuming a row-major
  // storage order. The data is fully copied.
  Eigen::MatrixXf auxMat(mat.rows(), mat.cols());
  auxMat = mat.transpose();
  std::vector<float> vec(auxMat.data(), auxMat.data() + auxMat.size());
  return vec;
}


Eigen::MatrixXf vectorToEigen(std::vector<float> vec) {
  // Converts std::vector<float> to Eigen::MatriXf assuming a row-major
  // storage order, i.e., [x0, y0, z0, x1, ..., zn] gets converted to
  // [[x0, y0, z0], [x1, y1, z1], ...] as an Eigen matrix. Data is fully
  // copied (deep copy).
  int j = 0;
  int i = 0;
  int nPts;

  Eigen::MatrixXf mat;
  nPts = ((int) vec.size() / DIM);
  mat = Eigen::MatrixXf::Zero(nPts, DIM);

  for(int k = 0; k < vec.size(); ++k) {
    j = k % DIM;
    i = ((int) k / DIM);
    mat(i, j) = vec[k];
  }
  return mat;
}


float degreesToRadians(float d) {
  // Converts an input in degrees to radians
  return (d / 180.0) * ((float) M_PI);
};


Eigen::Matrix3f rotMat(float alpha_deg, float beta_deg, float gamma_deg) {
  // Creates a 3x3 rotation matrix according to the provided rotation angles
  // about the z-axis, y-axis and x-axis (according to the order of the arguments)
  Eigen::Matrix3f Rx;
  Eigen::Matrix3f Ry;
  Eigen::Matrix3f Rz;
  float alpha_rad = degreesToRadians(alpha_deg);
  float beta_rad = degreesToRadians(beta_deg);
  float gamma_rad = degreesToRadians(gamma_deg);

  Rx << 1, 0, 0,
        0, cos(gamma_rad), sin(gamma_rad),
        0, sin(gamma_rad), cos(gamma_rad);

  Ry << cos(beta_rad), 0, sin(beta_rad),
        0, 1, 0,
        -sin(beta_rad), 0, cos(beta_rad);

  Rz << cos(alpha_rad), -sin(alpha_rad), 0,
        sin(alpha_rad), cos(alpha_rad), 0,
        0, 0, 1;

  Eigen::Matrix3f R = Rz*Ry*Rx;

  return R;
};


Eigen::MatrixXf srtWarp(Eigen::MatrixXf A, srtTransformation transf) {
  // Transforms a nPts x 3 point matrix A according to a SRT transformation
  Eigen::MatrixXf srtA;
  Eigen::MatrixXf transfA;
  srtA = transf.s * transf.R * A.transpose();
  srtA.colwise() += transf.t;
  transfA = srtA.transpose();
  return transfA;
};


srtTransformation procrustes(Eigen::MatrixXf A, Eigen::MatrixXf B) {
  // Solves the least squares Procrustes problem of finding the SRT
  // srtTransformation that minimizes the error E = sAR + tj - B
  srtTransformation transf = {};

  Eigen::Vector3f centroid_A = A.colwise().mean();
  Eigen::Vector3f centroid_B = B.colwise().mean();

  A.rowwise() -= centroid_A.transpose();
  B.rowwise() -= centroid_B.transpose();

  Eigen::MatrixXf C = A.transpose() * B;

  Eigen::JacobiSVD<Eigen::MatrixXf> svd(C, Eigen::ComputeThinU | Eigen::ComputeThinV);
  Eigen::Matrix3f U = svd.matrixU();
  Eigen::Matrix3f V = svd.matrixV();

  transf.R = V * U.transpose();

  if (transf.R.determinant() < 0) {
    Eigen::Matrix3f reflex;
    reflex << 1, 0, 0,
              0, 1, 0,
              0, 0, transf.R.determinant();
    transf.R = U * reflex * V.transpose();
  };

  transf.t = centroid_B - transf.R * centroid_A;
  transf.s = 1;

  return transf;
};


ANNpointArray EigenMatrixToANNpointArray(Eigen::MatrixXf cloud) {
  // Converts an EigenMatrix to an ANNpointArray to be used when
  // computing nearest neighbors.
  int nPts = cloud.rows();
  int dim = cloud.cols();
  ANNpointArray dataPts;
  dataPts = annAllocPts(nPts, dim);
  for(int i = 0; i < nPts; i++){
    ANNpoint p = annAllocPt(dim);
    p[0] = ((float) cloud(i,0));
    p[1] = ((float) cloud(i,1));
    p[2] = ((float) cloud(i,2));
    dataPts[i] = p;
  };
  return dataPts;
};


Eigen::MatrixXf* nearestNeighbors(Eigen::MatrixXf* clouds, int nClouds) {
  // Creates nClouds correspondence matrices with the indices of the
  // nearest neighbors of each cloud, in all the other clouds.
  int nPts = clouds[0].rows();
  int dim = clouds[0].cols();
  Eigen::MatrixXf* corr;
  corr = new Eigen::MatrixXf [nClouds];           // correspondence matrix

  // Initialize the correspondence matrices
  for(int i = 0; i < nClouds; i++) {
    corr[i] = Eigen::MatrixXf::Zero(nPts, nClouds);
  }

  // Allocate stuff before the search loop
  ANNidxArray nnIdx = new ANNidx[1];					    // near neighbor index
  ANNdistArray dists = new ANNdist[1];            // near neighbor distance
  ANNpoint queryPt = annAllocPt(dim);             // point we lookin for
  ANNpointArray dataPts = annAllocPts(nPts, dim); // search space

  for (int ki = 0; ki < nClouds; ki++) {
    ANNkd_tree* kdTree;
    dataPts = EigenMatrixToANNpointArray(clouds[ki]);
    kdTree = new ANNkd_tree(dataPts, nPts, dim);  // k-d tree init

    for (int kj = 0; kj < nClouds; kj++) {
      for (int i = 0; i < nPts; i++) {
          queryPt[0] = clouds[kj](i,0);
          queryPt[1] = clouds[kj](i,1);
          queryPt[2] = clouds[kj](i,2);
          kdTree->annkSearch(queryPt, 1, nnIdx, dists, 0);
          corr[kj].operator()(i,ki) = ((int) nnIdx[0]);
      }
    }
    delete kdTree;                                // clean things up
    annClose();
  }
  delete [] nnIdx;			                          // clean things up
  delete [] dists;                                // clean things up
  return corr;
};


void findGroups(Eigen::MatrixXf* corr, int nClouds, groupmap& m) {
  // Finds groups of points which are mutually nearest neighbors. A map m holds
  // the information about the groups. The keys are the std::vectors describing
  // the group, and the values are the number of occurences of each group.
  int nPts = corr[0].rows();
  int ind_in_i;                        // cloud-i index of the nearest point
  int ind_in_j;                        // cloud-j index of the nearest point
  int nMutuals;                        // no. of mutually nearest nbrs in a group

  for (int i = 0; i < nClouds; i++) {
    for (int p = 0; p < nPts; p++) {

      std::vector<int> ind = {};       // mutually nearest neighbors indices
      nMutuals = 0;

      for (int j = 0; j < nClouds; j++) {
        ind_in_j = corr[i](p,j);
        ind_in_i = corr[j](ind_in_j, i);

        if (p == ind_in_i) {           // mutually nearest neighbor condition
          nMutuals++;
          ind.insert(ind.end(), ind_in_j);
        }
        else {
          ind.insert(ind.end(), -1);   // if no mutually nearest point, add -1
        };
      };

      if (nMutuals > 1)
        ++m[ind];                      // add to the map if there's at least 2

    };
  };
};


Centroid::Centroid(Eigen::MatrixXf* clouds, int nClouds, groupmap m) {
  int count;
  int cloudNumber;
  int groupNumber;
  std::vector<int> validClouds;
  std::vector<int> validIndexes;
  std::vector<int> group;
  Eigen::MatrixXf groupCentroid(1,DIM);

  centralClouds = new std::vector<Eigen::MatrixXf> [nClouds];
  filteredClouds = new std::vector<Eigen::MatrixXf> [nClouds];
  groupNumber = 0;

  // iterate over all the groups in the map
  for(std::map<std::vector<int>, int>::iterator iter = m.begin(); iter != m.end(); ++iter) {
    group = iter->first;           // get the group itself
    count = iter->second;          // get the number of occurences of this group
    cloudNumber = 0;
    validClouds = {};
    groupCentroid = Eigen::MatrixXf::Zero(1,DIM);

    // check if group occurences equals number of clouds in the group
    if (isGroupValid(group, count)) {
      // iterate over all the indices in the group
      for (std::vector<int>::const_iterator j = group.begin(); j != group.end(); ++j) {
        if (*j != -1) {
          validClouds.insert(validClouds.end(), cloudNumber);
          groupCentroid.row(0) += clouds[cloudNumber].row(*j);
          filteredClouds[cloudNumber].insert(filteredClouds[cloudNumber].end(), clouds[cloudNumber].row(*j));
        }
        ++cloudNumber;
      };

      groupCentroid = groupCentroid / validClouds.size();

      for (std::vector<int>::const_iterator k = validClouds.begin(); k != validClouds.end(); ++k)
        centralClouds[*k].insert(centralClouds[*k].end(), groupCentroid);
    }
    ++groupNumber;
  };
};


bool Centroid::isGroupValid(std::vector<int> group, int count) {
  int numCloudsInGroup;
  // Get number of clouds in the group (number of entries different from -1)
  numCloudsInGroup = group.size() - std::count(group.begin(), group.end(), -1);

  if (numCloudsInGroup == count) {
    return true;
  }
  else {
    return false;
  }
};


Eigen::MatrixXf Centroid::centroidByCloudNum(int cloudNumber) {
  // Get a matrix with the centroid points calculated with cloud cloudsNumber
  Eigen::MatrixXf mat = Eigen::MatrixXf::Zero(centralClouds[cloudNumber].size(), DIM);
  float* pt;

  for (int i = 0; i < centralClouds[cloudNumber].size(); i++) {
    pt = centralClouds[cloudNumber][i].data();
    Eigen::MatrixXf mat_row = Eigen::Map<Eigen::Matrix<float, 1, 3>>(pt);
    mat.row(i) += mat_row;
  }
  return mat;
};


Eigen::MatrixXf Centroid::cloudByCloudNum(int cloudNumber) {
  // Get a matrix with the points from cloud cloudNumber which were used to
  // compute the centroid
  Eigen::MatrixXf mat = Eigen::MatrixXf::Zero(filteredClouds[cloudNumber].size(), DIM);
  float* pt;

  for (int i = 0; i < filteredClouds[cloudNumber].size(); i++) {
    pt = filteredClouds[cloudNumber][i].data();
    Eigen::MatrixXf mat_row = Eigen::Map<Eigen::Matrix<float,1,DIM> >(pt);
    mat.row(i) += mat_row;
  }
  return mat;
};


void gpaipc(Eigen::MatrixXf* clouds, int nClouds, int maxIter, bool verbose) {
  float meanSquaredError = 0;
  srtTransformation srt;
  Eigen::MatrixXf* indexMatrix;

  for (int iiter = 0; iiter < maxIter; ++iiter) {
    if (verbose == true)
      std::cout << "Epoch: "  << iiter << " \n";

    indexMatrix = nearestNeighbors(clouds, nClouds);
    groupmap groups;
    findGroups(indexMatrix, nClouds, groups);
    delete [] indexMatrix;
    Centroid centroid(clouds, nClouds, groups);

    Eigen::MatrixXf current;
    Eigen::MatrixXf target;

    // run procrustes for all the clouds with the centroid
    for (int i = 0; i < nClouds; ++i) {
      target = centroid.centroidByCloudNum(i);
      current = centroid.cloudByCloudNum(i);
      meanSquaredError = eigenMSE(target, current);

      if (verbose == true) {
        std::cout << " Loss (MSE) (cloud" << i << "): "  << meanSquaredError;
        std::cout << " (using " << target.rows() << " mutual neighbors)" << '\n';
      }

      srt = procrustes(current, target);
      clouds[i] = srtWarp(clouds[i], srt);
    }
  }
};
